users/models.py

```diff
@@ -601,3 +596,80 @@ class BlockList(TimestampedModel):
     """A user's blocklist model"""
 
     hashed_email = models.CharField(max_length=128)
+
+
+class UserOrganization(models.Model):
+    """The user's organizations memberships."""
+
+    user = models.ForeignKey(
+        User, on_delete=models.CASCADE, related_name="b2b_organizations"
+    )
+    organization = models.ForeignKey(
+        "b2b.OrganizationPage",
+        on_delete=models.CASCADE,
+        related_name="organization_users",
+    )
+    keep_until_seen = models.BooleanField(
+        default=False,
+        help_text="If True, the user will be kept in the organization until the organization is seen in their SSO data.",
+    )
+
+    class Meta:
+        unique_together = ("user", "organization")
+
+    def __str__(self):
+        """Return a reasonable representation of the object as a string."""
+
+        return f"UserOrganization: {self.user} in {self.organization}"
+
+    @staticmethod
+    def process_add_membership(user, organization, *, keep_until_seen=False):
+        """
+        Add a user to an org, and kick off contract processing.
+
+        This allows us to manage UserOrganization records without necessarily
+        being forced to process contract memberships at the same time.
+
+        Args:
+        - user (users.models.User): the user to add
+        - organization (b2b.models.OrganizationPage): the organization to add the user to
+        - keep_until_seen (bool): if True, the user will be kept in the org until the
+          organization is seen in their SSO data.
+        """
+
+        obj, created = UserOrganization.objects.get_or_create(
+            user=user,
+            organization=organization,
+        )
+        if created:
+            obj.keep_until_seen = keep_until_seen
+            obj.save()
+            try:
+                organization.attach_user(user)
+            except ConnectionError:
+                log.exception(
+                    "Could not attach %s to Keycloak org for %s", user, organization
+                )
+            organization.add_user_contracts(user)
+
+        return obj
+
+    @staticmethod
+    def process_remove_membership(user, organization):
+        """
+        Remove a user from an org, and kick off contract processing.
+
+        Other side of the process_add_membership function - removes the membership
+        and associated managed contracts.
+
+        Args:
+        - user (users.models.User): the user to remove
+        - organization (b2b.models.OrganizationPage): the organization to remove the user from
+        """
+
+        organization.remove_user_contracts(user)
+
+        UserOrganization.objects.filter(
+            user=user,
+            organization=organization,
+        ).get().delete()
```

Comment:
These methods don't access `cls` or `self` so to follow the pattern we typically use these should live in `b2b/api.py` instead